engines/e_pkcs11.c

```diff
@@ -0,0 +1,771 @@
+    if (rv != CKR_USER_NOT_LOGGED_IN && rv != CKR_OK) {
+        PKCS11_trace("C_Logout failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_LOGOUT, PKCS11_R_LOGOUT_FAILED);
+        goto err;
+    }
+    return 1;
+
+ err:
+    return 0;
+}
+
+static void pkcs11_end_session(CK_SESSION_HANDLE session)
+{
+    pkcs11_funcs->C_CloseSession(session);
+}
+
+CK_OBJECT_HANDLE pkcs11_get_private_key(PKCS11_CTX *ctx)
+{
+    CK_RV rv;
+    CK_OBJECT_CLASS key_class = CKO_PRIVATE_KEY;
+    CK_KEY_TYPE key_type = CKK_RSA;
+    CK_OBJECT_HANDLE objhandle;
+    size_t len_kc = sizeof(key_class), len_kt = sizeof(key_type), id_len;
+    unsigned long count;
+    CK_ATTRIBUTE tmpl[3];
+
+    id_len = strlen((char *)ctx->id);
+    tmpl[0].type = CKA_CLASS;
+    tmpl[0].pValue = &key_class;
+    tmpl[0].ulValueLen = len_kc;
+    tmpl[1].type = CKA_KEY_TYPE;
+    tmpl[1].pValue = &key_type;
+    tmpl[1].ulValueLen = len_kt;
+    tmpl[2].type = CKA_ID;
+    tmpl[2].pValue = ctx->id;
+    tmpl[2].ulValueLen = id_len;
+
+    rv = pkcs11_funcs->C_FindObjectsInit(ctx->session, tmpl,
+                                         sizeof(tmpl) / sizeof(CK_ATTRIBUTE) );
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsInit failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_INIT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjects(ctx->session, &objhandle, 1, &count);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjects failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+    return objhandle;
+
+ err:
+    return 0;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    size_t vlen, i, j = 0;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    *value = OPENSSL_malloc(strlen(tmp) + 1);
+
+    if (*value == NULL) goto err;
+
+    end = strpbrk(tmp, ";");
+
+    BIO_snprintf(*value, end == NULL ? strlen(tmp) + 1 :
+             (size_t) (end - tmp + 1), "%s", tmp);
+    hex2bin = OPENSSL_malloc(strlen(*value) + 1);
+
+    if (hex2bin == NULL) goto err;
+
+    vlen = strlen(*value);
+    for (i = 0; i < vlen; i++) {
+        if (*(*value+i) == '%' && i < (strlen(*value)-2)) {
+            *(hex2bin+j) = pkcs11_hex_int(*(*value+i+1), *(*value+i+2));
+            i += 2;
+        } else {
+            *(hex2bin+j) = *(*value+i);
+        }
+        j++;
+    }
+    *(hex2bin+j) = '\0';
+    *value = hex2bin;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static int pkcs11_get_console_pin(char **pin)
+{
+    int i, ret = 0;
```

Comment:
variable `i` is unused outside the  `#ifndef OPENSSL_NO_UI_CONSOLE ` block, so it will break compilation with no-ui.